### ISSUE TITLE
Bug: Visitors should receive all arguments from `VisitNodeStep.args`

### ISSUE THREAD

**Issue by jaymarvelz** [eslint/eslint#20261]
### Environment

Node version: v24.11.0
npm version: v11.6.2
Local ESLint version: v9.38.0
Global ESLint version: Not found
Operating System:

### What parser are you using?

Default (Espree)

### What did you do?

```js
import { ESLint } from "eslint";

const eslint = new ESLint({
  overrideConfigFile: true,
  overrideConfig: {
    plugins: {
      custom: {
        rules: {
          rule: {
            create() {
              return {
                Identifier(node, parent) {
                  console.log(parent.type);
                },
              };
            },
          },
        },
      },
    },
    rules: {
      "custom/rule": 2,
    },
  },
});

eslint.lintText("Foo");
```

### What did you expect to happen?

It should print `ExpressionStatement`. The visitor should receive both the `node` and its `parent` as defined in `VisitNodeStep.args`.

### What actually happened?

```
TypeError: Cannot read properties of undefined (reading 'type')
```

### Link to Minimal Reproducible Example

See above

### Participation

- [x] I am willing to submit a pull request for this issue.

**Comment by DMartens** [comment]
I think this is working as intended as only the ["general" analyzer](https://github.com/DMartens/eslint/blob/eac71fb77113de7bf199ff20c6ee44cefcb59848/lib/languages/js/source-code/source-code.js#L1300) sets the parent argument while the ["JavaScript" analyzer](https://github.com/DMartens/eslint/blob/eac71fb77113de7bf199ff20c6ee44cefcb59848/lib/languages/js/source-code/source-code.js#L1339) does not.
JavaScript is unique in this way as the `parent` property is set on each node while the other languages (Markdown, CSS and JSON) do not have the `parent` property, so the additional parameter for the visitor makes sense.

**Comment by crimsonjay0** [comment]
On my [branch](https://github.com/jaymarvelz/eslint/tree/forward-step-args), it prints `ExpressionStatement`, so I’m not sure what you mean by 'general analyzer' versus 'JavaScript analyzer'.

**Comment by mdjermanovic** [comment]
I think this is a bug (unintentional omission while implementing the language plugins design). The [Language Plugins RFC](https://github.com/eslint/rfcs/blob/40a07751123998d6437f4b9b1ee311fde7e79089/designs/2022-languages/README.md?plain=1#L308) defines property `args` on `VisitTraversalStep`, and it exists in the [`VisitTraversalStep` type](https://github.com/eslint/rewrite/blob/35f73425f5234755908debf26dc7e7353ffb2902/packages/core/src/types.ts#L1509) but is unused in ESLint. Linter passes just `VisitTraversalStep#target` as arguments:

Enter: https://github.com/eslint/eslint/blob/c7d32298482752eeac9fb46378d4f1ea095f3836/lib/linter/source-code-traverser.js#L291
Exit: https://github.com/eslint/eslint/blob/c7d32298482752eeac9fb46378d4f1ea095f3836/lib/linter/source-code-traverser.js#L303

@nzakas what do you think?

**Comment by nzakas** [comment]
Yup, this looks like an implementation bug. I'm kind of surprised we didn't catch it before this point, but I guess that means most people weren't using the second argument. :smile:
